Centralize daemon metrics

vvoland opened the pull request:
Centralize daemon metrics in an `internal/metrics` package.

thaJeztah commented on daemon/images/metrics.go:
```diff
@@ -1,18 +0,0 @@
-var ImageActions metrics.LabeledTimer
-
-func init() {
-	ns := metrics.NewNamespace("engine", "daemon", nil)
-	ImageActions = ns.NewLabeledTimer("image_actions", "The number of seconds it takes to process each image action", "action")
-	// TODO: is it OK to register a namespace with the same name? Or does this
-	// need to be exported from somewhere?
-	metrics.Register(ns)
```
Went looking if the `metrics.Register` was still needed, but look like this one is also registered in `daemon/metrics.go`;

https://github.com/moby/moby/blob/a72026acbbdfcf90f0ba203abd4e0943e3d546e7/daemon/metrics.go#L19
https://github.com/moby/moby/blob/a72026acbbdfcf90f0ba203abd4e0943e3d546e7/daemon/metrics.go#L57

vvoland replied:
Yeah, it's not needed because the namespace is already registered

thaJeztah commented on daemon/containerd/fake_service_test.go:
```diff
@@ -288,3 +290,8 @@ func (s *memoryLabelStore) Update(dgst digest.Digest, update map[string]string)
 	return labels, nil
 }
+
+var testImageActions = sync.OnceValue(func() metrics.LabeledTimer {
+	ns := metrics.NewNamespace("test", "containerd", nil)
+	return ns.NewLabeledTimer("image_action", "Image actions", "action")
+})
```
I started to wonder; how much sense would it make to move all those timers to a  `metrics` package, that would have some `sync.OnceValue` to get them? (instead of making these a field in the daemon).

Not sure if it's better, or if it would be _worse_. Mostly looking that some of the initialization is a bit sprinkled throughout, which means that some namespaces are registered multiple times in different locations, e.g. there's still one in `daemon/events`; https://github.com/moby/moby/blob/a72026acbbdfcf90f0ba203abd4e0943e3d546e7/daemon/events/metrics.go#L10-L15


But I guess that won't work for this specific case if we want separate counters for testing

vvoland committed:
Alias github.com/docker/go-metrics imports

Next commits will introduce a new internal `metrics` package, so alias
the "external" import to avoid confusion.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
daemon/metrics: Move out to `internal/metrics`

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
daemon/images: Move ImageActions to metrics

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
daemon: Move direct usages of go-metrics to internal/metrics

Add wrapper for StartTimer inside the internal package

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
daemon/events: Move metrics to internal/metrics

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

thaJeztah commented:
Looks like some file may need a build-tag;


```
#38 0.510 + go build -mod=vendor -modfile=vendor.mod -o /tmp/bundles/binary-daemon/dockerd.exe -tags ' osusergo static_build ' -ldflags '-w -X "github.com/docker/docker/dockerversion.Version=pr-49165" -X "github.com/docker/docker/dockerversion.GitCommit=1d17fa1b4f3780043bae865643e9c68afa83e17d" -X "github.com/docker/docker/dockerversion.BuildTime=2024-12-23T15:34:45.000000000+00:00" -X "github.com/docker/docker/dockerversion.PlatformName=Moby Engine - Nightly" -X "github.com/docker/docker/dockerversion.ProductName=moby-bin" -X "github.com/docker/docker/dockerversion.DefaultProductLicense="  -extldflags -static ' -gcflags= github.com/docker/docker/cmd/dockerd
#38 23.99 # github.com/docker/docker/internal/metrics
#38 23.99 internal/metrics/plugin.go:110:7: undefined: unix.Unlink
```

thaJeztah commented:
One more linter failure 😅 

```
daemon/daemon.go:126:2: field `metricsPluginListener` is unused (unused)
	metricsPluginListener net.Listener
	^
```

laurazard reviewed (approved):
LGTM, thanks! This is much better.

thaJeztah commented on daemon/container_operations.go:
```diff
@@ -453,7 +454,7 @@ func (daemon *Daemon) initializeNetworking(ctx context.Context, cfg *config.Conf
 		return nil, nil
 	}
 
-	// Cleanup any stale sandbox left over due to ungraceful daemon shutdown
+	// CleanupPlugin any stale sandbox left over due to ungraceful daemon shutdown
```
Oh; probably find/replace was too eager here

thaJeztah commented on internal/metrics/metrics.go:
```diff
@@ -0,0 +1,130 @@
+package metrics
```
For a follow-up; depending where this all ends up in (i.e. if it's also included in packages that are imported externally), we could even consider having a build-tag to disable metrics at compile time.

thaJeztah reviewed (approved):
LGTM

found one find/replace issue; not a blocker for sure, but in case you want to fix it up ❤️
